lib/CL/pocl_llvm_api.cc

```diff
@@ -1529,8 +1556,25 @@ static PassManager& kernel_compiler_passes
 	  passes.push_back("automatic-locals");
   if (SPMDDevice)
     passes.push_back("flatten-inline-all");
-  else
-    passes.push_back("flatten-globals");
+  else {
+    /* LLVM is unfortunately still not intelligent enough to flatten.
```

Comment:
Yes, unfortunately. Or perhaps we're missing some pass. I've noticed this in NBody (IIRC), it had a not inlined _cl_sqrt() or some such function with only a half dozen instructions. Had a relatively big negative impact on NBody's performance.